src/metaswitch/clearwater/etcd_shared/common_etcd_synchronizer.py

```diff
@@ -120,6 +122,129 @@ def _error_catcher(self): # pragma: no cover
+
+        if timeout == 0:
+            timeout = None
+
+        if not path.startswith('/'):
+            raise ValueError('Path does not start with /')
+
+        while not response:
+            try:
+                response = payload(self, path, method,
+                                   params=params, timeout=timeout)
+                # Check the cluster ID hasn't changed under us.  We use
+                # preload_content=False above so we can read the headers
+                # before we wait for the content of a watch.
+                self._check_cluster_id(response)
+                # Now force the data to be preloaded in order to trigger any
+                # IO-related errors in this method rather than when we try to
+                # access it later.
+                _ = response.data
+                # urllib3 doesn't wrap all httplib exceptions and earlier versions
+                # don't wrap socket errors either.
+            except (HTTPError, MaxRetryError, HTTPException, SocketError) as e:
+                if (isinstance(params, dict) and
+                    params.get("wait") == "true" and
+                    (isinstance(e,
+                                urllib3.exceptions.MaxRetryError) or
+                     isinstance(e,
+                                urllib3.exceptions.ReadTimeoutError))):
+                    _log.debug("Watch timed out.")
+                    raise etcd.EtcdWatchTimedOut(
+                        "Watch timed out: %r" % e,
+                        cause=e
+                    )
+                _log.error("Request to server %s failed: %r",
+                           self._base_uri, e)
+
+                if self._allow_reconnect:
+                    _log.info("Reconnection allowed, looking for another "
+                              "server.")
+                    # _next_server() raises EtcdException if there are no
+                    # machines left to try, breaking out of the loop.
+                    self._base_uri = self._next_server(cause=e)
+                    some_request_failed = True
+                    # if exception is raised on _ = response.data
+                    # the condition for while loop will be False
+                    # but we should retry
+                    response = False
+                else:
+                    _log.debug("Reconnection disabled, giving up.")
+                    raise etcd.EtcdConnectionFailed(
+                        "Connection to etcd failed due to %r" % e,
+                        cause=e
+                    )
+            except etcd.EtcdClusterIdChanged as e:
+                _log.warning(e)
+                raise
+            except:
+                _log.exception("Unexpected request failure, re-raising.")
+                raise
+
+            if some_request_failed:
+                if not self._use_proxies:
+                    # The cluster may have changed since last invocation
```

Comment:
As someone not very familiar with this code, this comment doesn't tell me very much (maybe it makes more sense in context though)